### ISSUE TITLE
Large model fails with "Aborted (core dumped)" on Linux, works fine on Windows and Ubuntu on WSL

### ISSUE THREAD

**Issue by jemiller0** [dotnet/efcore#10670]
I have a legacy database with a large number of tables which I generated an EF Core DbContext and entities for using my own custom code generator. The code is contained in a .NET Standard 2 library. I have a .NET Core 2 console application with a simple query in it that I am using to test the library. I am able to run the .NET Core console application without error on Windows as well as on Ubuntu Linux running on WSL (Windows Subsystem for Linux). If I run it on Red Hat 7 or Ubuntu running on separate hardware I receive an error of "Aborted (core dumped)" with no stack trace.

I have also had problems getting this large model to work in a Web Forms application running on Windows. In that case, the same code was contained in a .NET Framework 4.7.1 library. I think the problem was that EF Core was using a lot of stack space when initializing the model. A workaround that I found was to create a separate thread and explicitly set the stack size. Everything worked fine after that. I.e. I was able to use the DbContext on other threads without a problem. Also, I had a .NET Framework 4.7.1 console application and that always worked fine.

I attached a Visual Studio solution with the library and console application that demonstrates the problem to this issue.

[Ole7.zip](https://github.com/aspnet/EntityFrameworkCore/files/1616889/Ole7.zip)

I suspect that the problem on Linux is that it is running out of stack space. However, I don't know how to confirm that. The code includes a global catch block that prints a stack trace, but, nothing is being printed out. I'm not sure why it works on Ubuntu running on the Windows kernel, but, not on Linux.

Also, I tried it with the Pomelo provider and also the SQL Server provider and the behavior was the same.

The model includes 1,310 entities. I was able to get it to work if I reduced the number entities. I started adding them back and it failed at around 300 entities.

Is there something that could be done to reduce the amount of memory EF Core uses while initializing?

.NET Core: 2.1.3 on Ubuntu, 2.0.3 on Red Hat
EF Core version: 2.0.1
Database Provider: Pomelo.EntityFrameworkCore.MySql 2.0.1, Microsoft.EntityFrameworkCore.SqlServer 2.0.1
Operating system: Red Had Linux 7.4, Ubuntu Linux 17.04
IDE: Visual Studio 2017 15.5.3

**Comment by AndriySvyryd** [comment]
@jemiller0 I was able to repro on Windows by limiting stack size:
```C#
var thread = new Thread(Run, 100000);
thread.Start();
```

The StackOverflowException is thrown when `OleContext.OnModelCreating` is *about* to get executed. And it's simple because it's so gigantic. I wasn't able to confirm this using the debugger, but I suspect that all the lambda delegates created for the configuration calls just don't fit on the stack. You can workaround this in several ways:

* Break up OnModelCreating into smaller methods, for example by using [IEntityTypeConfiguration](https://docs.microsoft.com/en-us/ef/core/what-is-new/#self-contained-type-configuration-for-code-first)
* Use the string overloads instead of the lambda ones, i.e. `Entity<PatronAddress>().HasOne("AddressSource")` instead of `Entity<PatronAddress>().HasOne(pa => pa.AddressSource)`

**Comment by jemiller0** [comment]
Thanks, I will try using strings instead of lambdas. Also, I will try by creating an initializing the model on a separate thread with a larger stack size. My complaint at this point is about the .NET Core runtime. Why is it not printing a stack trace? Also, it seems odd that the behavior is different depending on what OS it's running on.

Maybe using the strings will speed up the initialization also. EF has always been slow to initialize and continues to be with EF Core. Personally, I think it should be lazily initialized. NHibernate doesn't have the lag that EF does.

**Comment by jemiller0** [comment]
Also, what confused me was that it didn't fail part way through OnModelCreating(). I put print statements in there and they weren't executed. As you mentioned, it failed before OnModelCreating() which was a mystery to me.

**Comment by jemiller0** [comment]
I removed all the lambda functions and instead used typeof()/nameof() with strings instead. This improved startup time from 16 seconds to 12. However, it is still failing on Linux. It does work, if I initialize the context on another thread with a 1 MB stack.

[Ole7.zip](https://github.com/aspnet/EntityFrameworkCore/files/1629066/Ole7.zip)

It would be great if the attributes were enhanced to support composite keys. I would prefer not using the fluent API to begin with. The only reason I'm using it is for configuring composite keys/foreign keys and for specifying the cascade behavior.

I just checked. It is failing even before OnModelCreating() is being called. So, how could it have to do with what's in OnModelCreating()?
